simulator/src/testbench.js

```diff
@@ -1,3 +1,512 @@
+    `;
+    $('#testBench').dialog({
+        resizable: false,
+        width: 'auto',
+        buttons: [
+            {
+                text: 'Run All',
+                click() {
+                    const testJSON = $('#testJSON').val();
+                    runTestBench(JSON.parse(testJSON), globalScope, MODE.MODE_RUNALL, CONTEXT.CONTEXT_SIMULATOR);
+                },
+            },
+
+            {
+                text: 'Run Manually',
+                click() {
+                    const testJSON = $('#testJSON').val();
+                    $(this).dialog('close');
+                    runTestBench(JSON.parse(testJSON), globalScope, MODE.MODE_MANUAL, CONTEXT.CONTEXT_SIMULATOR);
+                },
+            },
+        ],
+    });
+
+    $('#testBench').empty();
+    $('#testBench').append(s);
+}
+
+/**
+ * Interface function to run testbench. Called by testbench prompt on simulator or assignments
+ * @param {Object} data - Object containing Test Data
+ * @param {Mode=} mode - To run all tests automatically or let user manually control tests
+ * @param {RunContext=} runContext - Whether simulator or Assignment called this function
+ * @param {Scope=} scope - the circuit
+ */
+export function runTestBench(data, scope = globalScope, mode = MODE.MODE_RUNALL, runContext = CONTEXT.CONTEXT_SIMULATOR) {
+    if (mode === MODE.MODE_MANUAL && runContext === CONTEXT.CONTEXT_ASSIGNMENTS) {
+        // Assignment does not support MANUAL mode
+        return;
+    }
+    const isValidData = validate(data, scope);
+    if (!isValidData.ok){
+        showError(`TestBench: ${isValidData.message}`);
+        return;
+    }
+
+    let results;
+    if (mode === MODE.MODE_RUNALL) results = runAll(data, scope);
+    else if (mode === MODE.MODE_MANUAL) {
+        runManual(data, scope);
+        return;
+    } else {
+        showError('TestBench: Bad Test Mode');
+        return;
+    }
+
+    console.log(results);
+}
+
+/**
+ * Run all the tests automatically. Called by runTestBench()
+ * @param {Object} data - Object containing Test Data
+ * @param {Scope=} scope - the circuit
+ */
+function runAll(data, scope) {
+    // Stop the clocks
+    // TestBench will now take over clock toggling
+    changeClockEnable(false);
+
+    const { inputs, outputs, reset } = bindIO(data, scope);
+    for (const group of data.groups) {
+        for (const output of group.outputs) output.results = [];
+        for (let case_i = 0; case_i < group.n; case_i++) {
+            // Set and propagate the inputs
+            setInputValues(inputs, group, case_i, scope);
+            // If sequential, trigger clock now
+            if (data.type === 'seq') tickClock(scope);
+            // Get output values
+            const caseResult = getOutputValues(data, outputs);
+            // Put the results in the data
+            for (const outName of caseResult.keys()) {
+                const output = group.outputs.find((dataOutput) => dataOutput.label === outName);
```

Comment:
`.find()` is a linear time complexity. This might need refactoring.

Reply:
But our sample size is pretty low for this to be a problem right. The output count is fairly low for this to be significant